action.php

```diff
@@ -1,341 +1,348 @@
 <?php
 
 // Define params.
-define( 'BP', dirname( __FILE__ ) );
-define( 'DS', DIRECTORY_SEPARATOR );
+if (!defined('BP')) define( 'BP', dirname( __FILE__ ) );
+
+if (!defined('DS')) define( 'DS', DIRECTORY_SEPARATOR );
 
 // Disable cache.
 header( 'Cache-Control: no-cache, must-revalidate' );
 header( 'Expires: Sat, 26 Jul 1997 05:00:00 GMT' );
 header( 'Pragma: No-Cache' );
 
-// Allow only posted data.
-if ( empty( $_POST ) ) { die( 'Must be post method.' ); }
-
-// @todo: clean up posted data
-if ( ! isset( $_POST['op'] ) ) { die( 'No action or API key set.' ); }
-$_POST['op'] = trim( $_POST['op'] );
+function smaily_admin_save() {
+	// Allow only posted data.
+	if ( empty( $_POST ) ) { die( 'Must be post method.' ); }
+
+	// Parse form data out of the serialization.
+	$form_data = array();
+	parse_str( $_POST['form_data'], $form_data );
+
+	// Validate posted operation.
+	if ( ! isset( $form_data['op'] ) ) { die( 'No action or API key set.' ); }
+	$form_data['op'] = ( in_array( $form_data['op'], array( 'validateApiKey', 'removeApiKey', 'resetForm', 'refreshAutoresp', 'save' ), true )
+		?  $form_data['op'] : '' );
```

Comment:
The switch statement makes sure no code is ran if 'op' is not valid. Is this code portion useless if the switch statement does the same thing (validation)?

Reply:
Switch statement does not execute any code, but after the switch statement there is logic to "refresh" the page. Does that need to run when switch statement does not?

It is not useless, because you can short circuit function, if `op` is not valid.

Reply:
I don't think a refresh is needed if no valid operation is posted.

Reply:
If refresh is not needed then you can just stop the execution (short circuit) right after `op` value has been normalized.

Short circuiting is a better alternative (more explicit and future proof) to handling different values (or state) of an variable.

For example, this is easier to understand:
```php
$acceptable = false;

// Stop further code execution, if user is not acceptable.
if ($acceptable === false) {
    die("This is not acceptable!");
}

$id = $this->request->get('id', NULL);
if (!empty($id)) {
    DB::update('foo', array('bar' => 'xyz'), array('id' => $id));
}
else {
    DB::insert('foo', array('bar' => 'abc'));
}

echo "OK";
```

than:
```php
$acceptable = false;

$id = $this->request->get('id', NULL);
if (!empty($id) and $acceptable === true) {
    DB::update('foo', array('bar' => 'xyz'), array('id' => $id));
}
elseif ($acceptable === true) {
    DB::insert('foo', array('bar' => 'abc'));
}

if ($acceptable === false) {
    echo "This is not acceptable!";
}
else {
    echo "OK";
}
```